src/ICS20Transfer.sol

```diff
@@ -57,78 +65,71 @@ contract ICS20Transfer is IIBCApp, IICS20Transfer, IICS20Errors, Ownable, Reentr
-        // TODO Emit error event
+        // Since this function mostly returns acks, also when it fails, the ics26router (the caller) will log the ack
         if (keccak256(abi.encodePacked(msg_.packet.version)) != keccak256(abi.encodePacked(ICS20Lib.ICS20_VERSION))) {
+            // TODO: Figure out if should actually error out, or if just error acking is enough
             return ICS20Lib.errorAck(abi.encodePacked("unexpected version: ", msg_.packet.version));
         }
 
-        ICS20Lib.PacketDataJSON memory packetData = ICS20Lib.unmarshalJSON(msg_.packet.data);
+        ICS20Lib.UnwrappedPacketData memory packetData = _unwrapReceivePacketData(msg_.packet);
+
         if (packetData.amount == 0) {
-            return ICS20Lib.errorAck(abi.encodePacked("invalid amount: 0"));
+            return ICS20Lib.errorAck("invalid amount: 0");
         }
 
         (address receiver, bool receiverConvertSuccess) = ICS20Lib.hexStringToAddress(packetData.receiver);
         if (!receiverConvertSuccess) {
             return ICS20Lib.errorAck(abi.encodePacked("invalid receiver: ", packetData.receiver));
         }
 
-        // TODO: Handle non-contract denoms (destination chain is not source)
-        bytes memory denomPrefix = ICS20Lib.getDenomPrefix(msg_.packet.sourcePort, msg_.packet.sourceChannel);
-        bytes memory denom = bytes(packetData.denom);
-        if (
-            denom.length >= denomPrefix.length
-                && ICS20Lib.equal(ICS20Lib.slice(denom, 0, denomPrefix.length), denomPrefix)
-        ) {
-            // sender chain is not the source, unescrow tokens
-            // TODO: Implement escrow balance tracking (#6)
-
-            string memory unprefixedDenom =
-                string(ICS20Lib.slice(denom, denomPrefix.length, denom.length - denomPrefix.length));
-            (address tokenContract, bool tokenContractConvertSuccess) = ICS20Lib.hexStringToAddress(unprefixedDenom);
-            if (!tokenContractConvertSuccess) {
-                return ICS20Lib.errorAck(abi.encodePacked("invalid token contract: ", unprefixedDenom));
-            }
-
-            IERC20(tokenContract).safeTransfer(receiver, packetData.amount);
-        } else {
-            // sender chain is the source, mint vouchers
-            // TODO: Implement escrow balance tracking (#6)
-            // TODO: Implement creating (new erc20 contracts), looking up and minting of vouchers
-            revert ICS20UnsupportedFeature("sender denom is source");
+        // TODO: Implement escrow balance tracking (#6)
+        if (packetData.originatorChainIsSource) {
+            // sender is source, so we mint vouchers
+            // NOTE: The unwrap function already created a new contract if it didn't exist already
+            IBCERC20(packetData.erc20Contract).mint(packetData.amount);
         }
 
+        // transfer the tokens to the receiver
+        IERC20(packetData.erc20Contract).safeTransfer(receiver, packetData.amount);
```

Comment:
Yes, that would probably make the most sense. The reason it is not done right now is because with safeTransfer, which is a Library, we can't do try/catch (because it is technically not an external call right here). There is a little "hack" we can do instead, by wrapping the transfer in an external function and calling it with this.wrappedSafeTransfer. If you prefer that I can go ahead and implement that :)

Reply:
How is this not an external call? It seems we are calling `packetData.erc20Contract`?

Reply:
I know, but since the safeTransfer is an actually a library wrapper, we are not the doing the external call here. The safeTransfer library is doing it :P So we don't get an external call to try/catch, nor a return value.

```
Error (2536): Try can only be used with external function calls and contract creation calls.
   --> src/ICS20Transfer.sol:121:13:
    |
121 |         try IERC20(packetData.erc20Contract).safeTransfer(receiver, packetData.amount) {
    |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```